### ISSUE TITLE
Nmap Windows installer: use App Paths registry key instead of system PATH variable

### ISSUE THREAD

**Issue by dmiller-nmap** [nmap/nmap#2349]
https://docs.microsoft.com/en-us/windows/win32/shell/app-registration

Nmap's Windows installer currently (by default) adds the Nmap install directory to the system PATH so that nmap.exe can be run from a command prompt. Windows 7 and later supports per-application paths that avoid issues such as Nmap's DLLs being found by other applications due to its position in the system PATH.

**Comment by fyodor** [comment]
This will be great as I experienced the problem today.  I had Nmap 7.91 installed and then tried to update to the upcoming 7.92 release.  But the install failed ("Error opening file for writing: C:\Program Files (x86\Nmap\libcrypto-1_1.dll") because it turns out that another application I had running (a cloud syncing app named Insync) had accidentally loaded this OpenSSL DLL from the Nmap directory, presumably because we added that dir to the search path.  The solution was to right-click on Insync in the system tray and quit it, then select "Retry" on the Nmap Setup error dialogue.  I've verified that this is not a new issue--it happens when running the 7.91 installer while Nmap is already installed too.

**Comment by nnposter** [comment]
I might be missing something but...

> Nmap's Windows installer currently (by default) adds the Nmap install directory to the system PATH so that nmap.exe can be run from a command prompt.

AFAIK, `cmd.exe` and `powershell.exe` do not use `App Paths` entries to locate executables. In other words, `App Paths` can be used for de-conflicting DLLs and launching applications (such as from the Run dialog) but they do not help with "*running [nmap.exe] from a command prompt*". Nmap would still have to be surfaced by some other means.

**Comment by dmiller-nmap** [comment]
@nnposter It appears it only works if you use the ShellExecute or ShellExecuteEx API functions, which are not used by cmd.exe or powershell.exe. However, the "start" command and the "Run..." dialog do use these. Zenmap uses Python's `subprocess.Popen` interface, which uses CreateProcess instead, so that would not be able to take advantage of the feature, either.

Some things we may wish to consider before choosing a resolution for this idea:
1. How important is it for users to be able to run nmap from the command-line without specifying the path?
2. Can we create wrapper commands for each of our intended in-path commands which then use ShellExecute to find the app via `App Paths`? We could then put these in a subdir of the install directory and add that to the PATH, which would avoid the DLL search path issue.

**Comment by dmiller-nmap** [comment]
I realize I didn't point out the actual problem we ran into that prompted this idea. When running the Nmap installer to upgrade an existing Nmap installation, we encountered an error due to being unable to delete the OpenSSL DLLs, which were in use by another process. Investigation turned up a service that was loading these DLLs from our install path because of [Windows default DLL search order](https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order). If we used App Paths instead, no other applications would find and use Nmap's DLLs accidentally.

**Comment by fyodor** [comment]
@dmiller-nmap - Your idea for the wrapper command directory sounds like a great one if it works and turns out easy to implement!

**Comment by dmiller-nmap** [comment]
Another alternative: if using App Paths, you can run a command like so:
* cmd.exe: `start /b /wait nmap --version`
* powershell.exe: `Start-Process -NoNewWindow -Wait nmap --version`

This could be added to a powershell profile as an alias, though of course you could also hard-code the absolute path there, too.

**Comment by nnposter** [comment]
> When running the Nmap installer to upgrade an existing Nmap installation, we encountered an error due to being unable to delete the OpenSSL DLLs, which were in use by another process.

This particular issue, specific to OpenSSL, would be IMHO best resolved by statically linking the libraries. This obviously mitigates the OpenSSL conflict but, more importantly, it also offers targeted crypto customization, which in turn remediates a long-standing issue of not being able to fingerprint and run scripts against targets with weak TLS/SSL (#511, #583, #1485, ...).

For this very reason I have been compiling my own custom-configured and patched OpenSSL on both Linux and Windows for the last few years.

**Comment by dmiller-nmap** [comment]
@nnposter We do [build our own OpenSSL](https://svn.nmap.org/nmap-mswin32-aux/OpenSSL-upgrade.txt), enabling weak crypto as much as possible within the configurations offered by the OpenSSL developers. Obviously, Nmap often wants to connect to very old or broken implementations that standard OpenSSL refuses to connect to, and we are limited by things like SSL 2 being stripped out of the codebase. However, building and distributing our own DLL instead of statically linking does allow the various Nmap-suite tools (Nmap, Ncat, and Nping) to share the library, reducing the size of each. This also allows users who have more particular needs to compile their own OpenSSL DLLs and replace those without needing to build Nmap itself.

The problem here was that another software package was loading OpenSSL DLLs from *Nmap's install directory* and keeping them locked so that Nmap could not be upgraded. This likely represents a bug in that software package, but we would like to be robust against the problem in the future as well as compliant with Microsoft's application standards as much as possible, which includes the App Paths registry entry.

**Comment by nnposter** [comment]
Yes, I do understand the original issue and I agree that the fault is likely in the setup of the other application.

**Comment by goarblt** [comment]
hello
https://www.evil.com/
